process/eng-practices/Exceptions-Haskell.md

```diff
@@ -0,0 +1,45 @@
+---
+tags:
+  - needs/review
+  - needs/writing
+---
+# Exceptions in Haskell
+
+In `cardano-wallet`, we currently use the following conventions for dealing with exceptions:
+
+1. Avoid partial functions.\
+    Example: Do not use `head`, but use a pattern match, or consider using functions from `Safe` module, such as `headMay`. Considering all corner cases does catch bugs.
+
+2. Pure code vs `IO` actions
+    * In pure code, use `error` instead of `undefined` if really necessary. In this case, please also add a `HasCallStack` constraint so that a backtrace can be shown when the impossible happens.
+    * In `IO`, use `throwIO` to throw a regular exception.
```

Comment:
I do like to recommend `UnliftIO.Exception.throwString` to lazy developers as a more user-friendly alternative to `error`.